broadinstitute/single_cell_portal_core
# Creating `ImagePipelineService` for running Image Pipeline jobs in PAPI (SCP-4649)

#### BACKGROUND
As part of ongoing Image Pipeline work, we will need a way to easily launch jobs in PAPI without having to manually construct parameter sets and poll for completion.  This will be a requirement for future work to automate the end-to-end run of generating expression arrays and then rendering static images for expression plots.

#### CHANGES
This update adds `ImagePipelineService` and `RenderExpressionArraysParameters`, both of which are modeled after the analogous `DifferentialExpression` classes.  These will allow launching of PAPI jobs through a simple convenience method of `ImagePipelineService.run_render_expression_arrays_job(study, cluster_file, matrix_file)` which will handle constructing/validating parameters, job submission/polling, and success/error reporting.  `RenderExpressionArraysParameters` will also dynamically scale the `machine_type` parameter based on the size of the expression matrix.

In addition, the `Parameterizable` concern was added to refactor out shared logic for constructing command line arrays for Docker/PAPI.  Lastly, this updates the version of `rubocop` and some of our linting rules to better support changes in Ruby 3.1

#### MANUAL TESTING
1. Pull branch, and enter a Rails console session:
```
./rails_local_setup.rb && source config/secrets/.source_env.bash && bin/rails c
```
2. In a separate terminal, start DelayedJob: 
```
./rails_local_setup.rb && source config/secrets/.source_env.bash && bin/rails jobs:work
```
3. Back in the rails console, load the `Murine vision and HIV with arrays` synthetic study (or any study with a small dense matrix & cluster file):
```
name = "Murine vision and HIV with arrays"
study = Study.find_by(name:)
cluster_file = study.cluster_ordinations_files.first
matrix_file = study.expression_matrices.first
```
4. Launch the job and confirm you get `true` back
```
ImagePipelineService.run_render_expression_arrays_job(study, cluster_file, matrix_file)
```
5. In `development.log`, note that the job has launched, and used the "small" machine type of `n1-highcpu-4`:
```
Remote found for cluster.tsv:6261a3fb94ec8fe535ac00a4, launching Ingest job
Request object sent to Google Pipelines API (PAPI), excluding 'environment' parameters:
---
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 6261a3e394ec8fe535ac009a
  - "--study-file-id"
  - 6261a3fb94ec8fe535ac00a4
  - "--user-metrics-uuid"
  - f775823b-94f7-46af-b1c7-789582efaa18
  - render_expression_arrays
 ...
  :virtual_machine:
    :boot_disk_size_gb: 300
    :machine_type: n1-highcpu-4
 ...
```
6. Back in the Rails console, get the link to the workspace bucket and open it in a browser:
```
study.google_bucket_url
```
7. Once the job completes, navigate to `_scp_internal/cache/expression_scatter/data/cluster_tsv` and validate that there are files there (if you used the synthetic study, you should see the following files):
![Screen Shot 2022-10-06 at 10 29 46 AM](https://user-images.githubusercontent.com/729968/194374008-39136928-06b1-4090-b3a8-101320f5efd1.png)
8. Click on any file, then click the link for "Authenticated URL", and validate the file transcodes in the browser and gives you an array of expression values
